File: examples/ext_io.cpp
#include "example_common.h"
#include "iroha/iroha.h"

using namespace iroha;

IDesign *build_design() {
  IDesign *design = new IDesign;

  IModule *module = new IModule(design, "M");
  design->modules_.push_back(module);
  ITable *table = new ITable(module);
  module->tables_.push_back(table);
  IState *st1 = new IState(table);
  IState *st2 = new IState(table);
  IState *st3 = new IState(table);
  table->states_.push_back(st1);
  table->states_.push_back(st2);
  table->states_.push_back(st3);

  IRegister *my_reg = DesignTool::AllocRegister(table, "my_reg", 32);

  IResource *ext_input = DesignTool::GetOneResource(table, resource::kExtInput);
  ResourceParams *params_input = ext_input->GetParams();
  params_input->SetExtInputPort("data_in", 32);
  IInsn *input_insn = new IInsn(ext_input);
  input_insn->outputs_.push_back(my_reg);
  st1->insns_.push_back(input_insn);

  IResource *ext_output = DesignTool::GetOneResource(table, resource::kExtOutput);
  ResourceParams *params_output = ext_output->GetParams();
  params_output->SetExtOutputPort("data_out", 32);
  IInsn *output_insn = new IInsn(ext_output);
  output_insn->inputs_.push_back(my_reg);
  st2->insns_.push_back(output_insn);

  table->SetInitialState(st1);
  DesignTool::AddNextState(st1, st2);
  DesignTool::AddNextState(st2, st3);
  DesignTool::Validate(design);

  return design;
}

File: examples/ext_io.py
import sys
sys.path.append('../py')

from iroha import *
from iroha.iroha import *

d = IDesign()
mod_top = IModule(d, "mod_top")

mod_sub = IModule(d, "mod_sub")
mod_sub.parent_module = mod_top
tab = ITable(mod_sub)
st1 = IState(tab)
st2 = IState(tab)
st3 = IState(tab)
tab.initialSt = st1
tab.states.append(st1)
tab.states.append(st2)
tab.states.append(st3)
design_tool.AddNextState(st1, st2)
design_tool.AddNextState(st2, st3)

ext_input = design_tool.CreateExtInput(tab, "data_in", 32)
iinsn = IInsn(ext_input)
r = IRegister(tab, "r")
iinsn.outputs.append(r)
st1.insns.append(iinsn)

ext_output = design_tool.CreateExtOutput(tab, "data_out", 32)
oinsn = IInsn(ext_output)
rc = design_tool.AllocConstNum(tab, False, 32, 123)
oinsn.inputs.append(rc)
st2.insns.append(oinsn)

ext_output_pulse = design_tool.CreateExtOutput(tab, "data_out_pulse", 1)
ext_output_pulse.resource_params.AddValue("DEFAULT-VALUE", "0")
oinsn_pulse = IInsn(ext_output_pulse)
rc = design_tool.AllocConstNum(tab, False, 32, 123)
oinsn_pulse.inputs.append(rc)
st2.insns.append(oinsn_pulse)

design_tool.ValidateIds(d)
w = DesignWriter(d)
w.Write()
